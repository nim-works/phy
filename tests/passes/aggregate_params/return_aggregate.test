discard """
  description: "
    Ensure that return aggregate values are turned into out parameters
    correctly, and that the callsites are properly transformed. The following
    cases need to be tested:
    * using an aggregate-returning call as an `Asgn` source
    * using an aggregate-returning call as a `Store` source
    * using an aggregate-returning call as a `Return` operand
  "
  output: "(Done)"
"""
(TypeDefs
  (Record 4 4
    (Field 0 (Int 4)))
  (ProcTy (Type 0))
  (ProcTy (Type 0) (Int 4))
  (ProcTy (Void)))
(GlobalDefs)
(ProcDefs
  (ProcDef (Type 2)
    (Locals (Int 4) (Type 0))
    (List
      (Block
        (Params (Local 0))
        (Return
          (Copy (Local 1))))))

  (ProcDef (Type 1) (Locals)
    (List
      (Block (Params)
        (Return
          (Call (Proc 0) (IntVal 100))))))

  (ProcDef (Type 3)
    (Locals (Type 0))
    (List
      (Block (Params)
        (Asgn (Local 0)
          (Call (Proc 0) (IntVal 100)))
        (Store (Type 0)
          (Addr (Local 0))
          (Call (Proc 0) (IntVal 100)))
        (Return)))))
