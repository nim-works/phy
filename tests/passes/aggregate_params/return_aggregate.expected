;$sexp
(TypeDefs
  (Record 4 4
    (Field 0 (Int 4)))
  (ProcTy (Void) (UInt 8))
  (ProcTy (Void) (Int 4) (UInt 8))
  (ProcTy (Void)))
(GlobalDefs)
(ProcDefs
  (ProcDef (Type 2)
    (Locals (Int 4) (Type 0) (UInt 8))
    (List
      (Block
        (Params (Local 0) (Local 2))
        (Store (Type 0)
          (Copy (Local 2))
          (Copy (Local 1)))
        (Return))))
  (ProcDef (Type 1)
    (Locals (UInt 8))
    (List
      (Block
        (Params (Local 0))
        (Call (Type 2) (Proc 0) (IntVal 100)
          (Copy (Local 0)))
        (Return))))
  (ProcDef (Type 3)
    (Locals (Type 0) (Type 0) (Type 0) (Type 0))
    (List
      (Block (Params)
        (Call (Type 2) (Proc 0) (IntVal 100)
          (Addr (Local 1)))
        (Asgn (Local 0)
          (Copy (Local 1)))
        (Call (Type 2) (Proc 0) (IntVal 100)
          (Addr (Local 2)))
        (Store (Type 0)
          (Addr (Local 0))
          (Copy (Local 2)))
        (CheckedCall (Type 2) (Proc 0) (IntVal 100)
          (Addr (Local 3))
          (Goto 1) (Unwind)))
      (Block (Params)
        (Asgn (Local 0)
          (Copy (Local 3)))
        (Return)))))
